var/spack/repos/builtin/packages/py-openslide-python/package.py

```diff
@@ -19,4 +19,4 @@ class PyOpenslidePython(PythonPackage):
     depends_on('openslide@3.4.0:')
     depends_on('python@2.6:2.8,3.3:')
     depends_on('py-setuptools', type='build')
-    depends_on('py-pillow+jpeg+jpeg2000+tiff', type=('build', 'run'))
+    depends_on('pil+jpeg+jpeg2000+tiff', type=('build', 'run'))
```

Comment:
This variant forwarding doesn't work yet. Should we just revert to `py-pillow` or `py-pillow-simd` instead of using a virtual dependency?

Reply:
Does this work?
```suggestion
    depends_on('pil')
    depends_on('py-pillow+jpeg+jpeg2000+tiff', type=('build', 'run'), when='^py-pillow')
    depends_on('py-pillow-simd+jpeg+jpeg2000+tiff', type=('build', 'run'), when='py-pillow')
```

Reply:
@becker33 that gives a very confusing error message:
```console
$ spack spec py-openslide-python
Input spec
--------------------------------
py-openslide-python

Concretized
--------------------------------
==> Error: '<' not supported between instances of 'NoneType' and 'str'
$ spack -d spec py-openslide-python
==> [2020-09-02-17:24:15.433058] Imported spec from built-in commands
==> [2020-09-02-17:24:15.436793] Imported spec from built-in commands
Input spec
--------------------------------
py-openslide-python

Concretized
--------------------------------
==> [2020-09-02-17:24:15.440299] Reading config file /Users/Adam/spack/etc/spack/defaults/config.yaml
==> [2020-09-02-17:24:15.485535] Reading config file /Users/Adam/spack/etc/spack/defaults/repos.yaml
==> [2020-09-02-17:24:16.054447] Reading config file /Users/Adam/spack/etc/spack/defaults/packages.yaml
==> [2020-09-02-17:24:16.086744] Reading config file /Users/Adam/spack/etc/spack/defaults/darwin/packages.yaml
==> [2020-09-02-17:24:16.112161] Reading config file /Users/Adam/.spack/darwin/packages.yaml
Traceback (most recent call last):
  File "/Users/Adam/spack/bin/spack", line 64, in <module>
    sys.exit(spack.main.main())
  File "/Users/Adam/spack/lib/spack/spack/main.py", line 762, in main
    return _invoke_command(command, parser, args, unknown)
  File "/Users/Adam/spack/lib/spack/spack/main.py", line 490, in _invoke_command
    return_val = command(parser, args)
  File "/Users/Adam/spack/lib/spack/spack/cmd/spec.py", line 103, in spec
    spec.concretize()
  File "/Users/Adam/spack/lib/spack/spack/spec.py", line 2211, in concretize
    self._expand_virtual_packages(concretizer),
  File "/Users/Adam/spack/lib/spack/spack/spec.py", line 2132, in _expand_virtual_packages
    copy.normalize(force=True)
  File "/Users/Adam/spack/lib/spack/spack/spec.py", line 2696, in normalize
    visited, all_spec_deps, provider_index, tests)
  File "/Users/Adam/spack/lib/spack/spack/spec.py", line 2634, in _normalize_helper
    dep, visited, spec_deps, provider_index, tests)
  File "/Users/Adam/spack/lib/spack/spack/spec.py", line 2536, in _merge_dependency
    provider = self._find_provider(dep, provider_index)
  File "/Users/Adam/spack/lib/spack/spack/spec.py", line 2471, in _find_provider
    providers = provider_index.providers_for(vdep)
  File "/Users/Adam/spack/lib/spack/spack/provider_index.py", line 78, in providers_for
    return sorted(s.copy() for s in result)
  File "/Users/Adam/spack/lib/spack/llnl/util/lang.py", line 231, in <lambda>
    lambda s, o: o is not None and s._cmp_key() < o._cmp_key())
TypeError: '<' not supported between instances of 'str' and 'NoneType'
```

Reply:
@adamjstewart Is that because of my typo where I put `py-pillow` instead of `^py-pillow-simd` in the second `when` clause, or did you fix that and then hit the error?

Reply:
Nah, I fixed that. Here is the actual diff I tested:
```diff
diff --git a/var/spack/repos/builtin/packages/py-openslide-python/package.py b/var/spack/repos/builtin/packages/py-openslide-python/package.py
index 286b07350a..6b10e62c2f 100644
--- a/var/spack/repos/builtin/packages/py-openslide-python/package.py
+++ b/var/spack/repos/builtin/packages/py-openslide-python/package.py
@@ -19,4 +19,6 @@ class PyOpenslidePython(PythonPackage):
     depends_on('openslide@3.4.0:')
     depends_on('python@2.6:2.8,3.3:')
     depends_on('py-setuptools', type='build')
-    depends_on('pil+jpeg+jpeg2000+tiff', type=('build', 'run'))
+    depends_on('pil', type=('build', 'run'))
+    depends_on('py-pillow+jpeg+jpeg2000+tiff', when='^py-pillow', type=('build', 'run'))
+    depends_on('py-pillow-simd+jpeg+jpeg2000+tiff', when='^py-pillow-simd', type=('build', 'run'))
```

Reply:
Also, even if that weird error didn't occur, I imagine that this would just result in the concretizer choosing `py-pillow~jpeg2000` and then raising a conflict error since the `depends_on` without `when` seem to be evaluated before the `depends_on` without `when`. Can't wait for that new concretizer...

Reply:
I think we should probably just leave this one depending on py-pillow for now, until we have a concretizer that can handle it.